### ISSUE TITLE
Auto-mount emptyDir volumes for `k8sapi/kubelet` or `runAsNotRoot: true`(?) executors

### ISSUE THREAD

**Issue by sachua** [argoproj/argo-workflows#2679]
# Summary

Will there be support for outputs from base image layer instead of using emptyDir to save params/artifacts when using the kubelet or k8s api executors?

# Motivation

I was running Kubeflow Pipelines with the container runtime executor as kubelet, but I kept hitting errors with the pipelines completing their tasks.

I then tried the https://github.com/argoproj/argo/blob/master/examples/artifact-passing.yaml example to try to narrow down the problem and realized that I had to mount an emptyDir volume in order for it to work.

My question is will there be support for outputs from base image layer with k8s api and kubelet executors in the future? Or is it not possible due to the way these executors work?

**Comment by alexec** [comment]
We use PNS ourselves - so this would be a great one for a community contribution.

**Comment by dtaniwaki** [comment]
As far as I know, it’s not possible now. 
https://github.com/kubernetes/kubectl/issues/454

**Comment by Ark-kun** [comment]
Argo already adds volumes to capture artifacts for docker executor. Why not for other executors? If Argo tells us to "mount an emptyDir volume", why can't it just auto-mount that volume itself?

**Comment by Ark-kun** [comment]
Our users need artifacts to work for some non-docker executor.

**Comment by sarabala1979** [comment]
I agree with @Ark-kun. The controller should add emptydir volume automatically if non-docker executors configured. This is a nice feature and will reduce many support questions.

**Comment by Ark-kun** [comment]
Related: https://github.com/kubeflow/pipelines/issues/1654 https://github.com/kubeflow/pipelines/issues/561 https://github.com/kubeflow/pipelines/issues/3407

**Comment by Ark-kun** [comment]
Let's keep this up

**Comment by nakfour** [comment]
Hey everyone, we use the "k8sapi" executor to run Argo on Openshift. Wanted to see if there is progress on this issue?

**Comment by alexec** [comment]
I've spent some time updating the documentation for the [executors](https://argoproj.github.io/argo/workflow-executors/) and how and when to use [emptyDir]https://argoproj.github.io/argo/empty-dir/).

TL;DR:

* Use the `docker` executor if you don't care about security and do care about performance. Use the `k8sapi` executor if you care about security and not about performance. Use PNS for something in between. Don't use `kubelet`.
* If you use `emptyDir` you don't get many of the security problems.

It is reasonable for the `kubelet` and `k8sapi` executors to automatically mount empty dir volumes for output artifacts.

To do this, copy `addInputArtifactsVolumes` and make the obvious changes.

Would anyone like to submit a PR?

**Comment by alexec** [comment]
> Argo already adds volumes to capture artifacts for docker executor.

@Ark-kun I'm not sure this is correct?

**Comment by Ark-kun** [comment]
> @Ark-kun I'm not sure this is correct?

AFAIK, Argo tries to use mirrored volumes to make the input artifacts accessible to the main container and capture the output artifacts.

See https://github.com/argoproj/argo/blob/4de64c618dea85334c0fa04a4dbc310629335c47/workflow/controller/workflowpod.go#L793 https://github.com/argoproj/argo/blob/4de64c618dea85334c0fa04a4dbc310629335c47/workflow/controller/workflowpod.go#L896 
https://github.com/argoproj/argo/blob/e14f4f922ff158b1fa1e0592fc072474e3257bd9/workflow/executor/executor.go#L369

**Comment by Ark-kun** [comment]
Perhaps we should add logic such that for every output artifact there should be an emptyDir mounted at it's parent directory unless there is another volume already mounted underneath.

**Comment by alexec** [comment]
thank you @Ark-kun

**Comment by Ark-kun** [comment]
I'm not sure this has been fixed.

**Comment by alexec** [comment]
Re-opened. This got closed by a PR and should not have been close. Would anyone be interested in investigating how to fix this?

**Comment by alexec** [comment]
This would be for `k8sapi` and `kubelet` executors. I think it may also be needed for whenever `runAsNotRoot: true`.

**Comment by alexec** [comment]
`argoproj/argoexec:auto-emptydir` and `argoproj/workflow-controller:auto-emptydir` available for anyone who wants to test this.

**Comment by alexec** [comment]
I'm going to wait for user testing before proceeding.

**Comment by alexec** [comment]
bump! please let me know if you really want this by testing the controller

**Comment by juliusvonkohout** [comment]
> bump! please let me know if you really want this by testing the controller

I think it is safe to say that we do really want this. Just have a look at the number of upvotes for such a technical issue and the other similar bugreports in kubeflow etc. It might help if  you could provide images for testing or even some short instructions for a kubernetes cluster. Then everyone can test his workloads. Otherwise everyone has to dig into how to build the images etc.

**Comment by alexec** [comment]
I think this issue will be moot with the Emissary Executor. 

[What is the Emissary Executor](https://github.com/argoproj/argo/blob/22f4d26ec059ad86b9c409ffe25392261a22afba/docs/workflow-executors.md#emissary-emissary)?

Why?

* As fast as the fastest executor (`docker`)
* As secure as the securest executor (`k8iapi`).
* Every feature of all other executors.

@terrytangyuan @Ark-kun would you be able to test this new executor, please?

You need dev builds of the controller and executor:

* `alexcollinsintuit/workflow-controller:emissary`
* `alexcollinsintuit/argoexec:emissary`

**Comment by juliusvonkohout** [comment]
I have tested it with kubeflow pipelines 1.3

There are only minor issues. 
1. op.set_security_context(V1SecurityContext(run_as_user=1000))
leads to permission errors in /var/argo/templates
But without that addition all pipelines finished successfully with runasnonroot.

2. Kubeflow pipelines runs with argo 2.7.5 by default. So the artifacts etc where successfully uploaded to minio. Nevertheless it seems that there was a communication error such that i cannot download them from the kubeflow pipelines webinterface anymore, even though they are in minio. So the output location is most likely not properly submitted. This is also the case if i use pns with your special emissary images, so i guess the 2.7.5 interface is just in some way not compatible with 2.12+ . 

Since kubeflow pipelines are going to use argo 2.12 in the future, it would be good if the emissary executor is compatible with 2.12.

**Comment by alexec** [comment]
I don't plan to backport the emissary to v2.12. It's a big and therefore risky change.

**Comment by alexec** [comment]
I think v2.12 is the right version for Kubeflow. I'd like us to get more involved and help out with Kubeflow more. Can you add me to the PR you're working on please?

**Comment by juliusvonkohout** [comment]
Well if the interfaces are somewhat compatible its fine. Then i just have to use the upstream argoexec and workflow-controller images for the emissary executor. That worked for 2.7.5 with 2.11 images as far as i remember.

https://github.com/kubeflow/pipelines/issues/4553 is the issue for argo 2.12 in kubeflow
and i think you are already in the conversation because of https://github.com/argoproj/argo/discussions/4953
